### ISSUE TITLE
Failed to load librefprop.so from cpp code in Linux

### ISSUE THREAD

**Issue by Yaykhel** [usnistgov/REFPROP-wrappers#335]
### Description
I was unable to load "librefprop.so" from the cpp code in Linux. The cpp code gives correct answers on Windows when using DLL file. The same code was tested on Linux using librefprop.so file. The librefprop.so file works fine when tested with ctREFPROP in python. It seems that the library file was not loaded successfully since this message "Loaded refprop (in main.cpp): false @ address 0" was displayed in the terminal.

"REFPROP_lib.h" and "REFPROP_underscore_lowercase_renaming.h" were included in the program.
 The program was complied as 

```
clang++-10 -lstdc++ -std=c++17  -ldl  -Ofast  REFPROP.cpp  objREFPROP.cpp  -o objREFPROP
```
If  -ldl  flag was not included, the error was "undefined reference to `dlsym'" 

Do we need modification of "REFPROP_lib.h", but it seems guards for various OSs were included in the file?

I looked into the example file given in "https://github.com/usnistgov/REFPROP-manager/blob/master/test/test.cpp" as well. The way of loading library is different.

It will be very helpful if a simple cpp progam for calling REFPROP linux is provided.

Please help with this problem.

Thank you.

### Steps to Reproduce

1. include REFPROP headers
2. bool loaded_REFPROP = load_REFPROP(err, path, librefprop.so);
3. complie and run

**Expected behavior:**
Successful loading of REFPROP

**Actual behavior:**
It seems the library file was not loaded.

### Versions

**REFPROP Version:** [REFPROP version 10]  
**Operating System and Version:** [Ubuntu 20.04.1]  
**Access Method:** [c++ wrapper]

### Additional Information

![LibraryLoadError](https://user-images.githubusercontent.com/45446429/97156266-d3848c80-17b9-11eb-88c7-4c2fe788d800.png)
![LibraryLoadFailed1](https://user-images.githubusercontent.com/45446429/97156275-d7181380-17b9-11eb-8e54-fac7bbf1184a.png)

**Comment by ianhbell** [comment]
Please post your simple example code and describe how you built the shared library.  Why did you include the other header file?  You only need the ``REFPROP_lib.h`` header file and to define the correct macros.  I think you have inadvertently shot yourself in the foot, memory-wise.  That seems clear to me since your ``ctREFPROP`` interfacing was a success.

**Comment by Yaykhel** [comment]
Thank you for your reply. I included REFPROP_lib.h only. I tested with the program you posted in cpp wrapper and it works. There seems to have string passing problem in my other program. Below is the program that prints out the correct values.

I built the shared library following your instructions and completed with two warnings as

```
Missing from PASS_CMN_tokens:  []
Missing from REFPROP_lib.h: []

```
Please let me know if these two lines are critical and how to solve it.

```
// Only this file gets the implementation

// clang++-10 -lstdc++ -std=c++17  mainREFPROP.cpp -ldl -o mainREFPROP

#define REFPROP_IMPLEMENTATION
#define REFPROP_FUNCTION_MODIFIER
#include "REFPROP_lib.h"
#undef REFPROP_FUNCTION_MODIFIER
#undef REFPROP_IMPLEMENTATION

#include <iostream>
#include <string>
#include <cstdlib>
#include <cstdio>

int main()
{
  // You may need to change this path to suit your installation
  // Note: forward-slashes are recommended.
  std::string path = "../cppLibraries/REFPROP";
  std::string DLL_name = "librefprop.so";

  // Load the shared library
  std::string err;

  bool loaded_REFPROP = load_REFPROP(err, path, DLL_name);
  printf("Loaded refprop (in main.cpp): %s @ address %zu\n", loaded_REFPROP ? "true" : "false", REFPROP_address());
  if (!loaded_REFPROP)
  {
    return EXIT_FAILURE;
  }

  SETPATHdll(const_cast<char *>(path.c_str()), 400);

  int ierr = 0, nc = 2;
  char herr[255], hfld[10000] = "AMMONIA|WATER", hhmx[255] = "HMX.BNC", href[4] = "DEF";
  SETUPdll(nc, hfld, hhmx, href, ierr, herr, 10000, 255, 3, 255);
  if (ierr > 0)
    printf("This ierr: %d herr: %s\n", ierr, herr);
  {
    int ierr = 0;
    char herr[255];
    double z[20] = {1.0}, x[20] = {1.0}, y[20] = {1.0}, T = 300, p = 101.325, d = -1, dl = -1, dv = -1, h = -1, s = -1, u = -1, cp = -1, cv = -1, q = -1, w = -1;
    TPFLSHdll(T, p, z, d, dl, dv, x, y, h, s, u, cp, cv, w, q, ierr, herr, 255);
    if (ierr > 0)
      printf("This ierr: %d herr: %s\n", ierr, herr);
    printf("This d: %g mol/L\n", d);
    {
      double T = 300;
      double z[20] = {0.2021, 0.7979};
      int kph = 1;
      double P, rhol, rhov;
      double x[20], y[20];
      int ierr;
      char herr[255];
      SATTdll(T, z, kph, P, rhol, rhov, x, y, ierr, herr, 255);
      printf("This P: %g kPa\n", P);
    }
  }

  return EXIT_SUCCESS;
}
```
The outputs are the same on both Windows and Linux.

The rest seems the cpp programming problem. I noticed that the path and dll name strings were jumping around when I print them out on Linux but not on Windows.

Thanks.

**Comment by ianhbell** [comment]
Yes, there are a few C++ issues in your code:

* You need to create a space-padded buffer in C++ of full length into which you place the string going into SETPATHdll.
* Use absolute paths, not relative ones in SETPATHdll

The second point is more a strong recommendation, the first is mandatory

**Comment by Yaykhel** [comment]
Thanks. I just used ../ not to show the directory. I use the absolute path in the running program.
I don't quite understand what is "need to create a space-padded buffer in C++ of full length". Please explain more. Do you mean space in front or in between or last?
I loaded a .txt file where the path, the refrigerant information and write directory were stored using getline() into std::vector<std::string> v. Then load path = v[i] and put it into load_REFPROP(). The scheme works perfectly on Windows while it failed in Linux. Instead I have to avoid the vector style and give the path string in "/dir" to make it work. I just don't know why.

**Comment by ianhbell** [comment]
You must *always* create string buffers that are large enough. REFPROP makes no guarantees that  it will only read up to the terminating NULL character of the C/C++ string, so if you have too short a buffer, all sorts of weird things can happen, and may well be system-dependent. Here is an example of how to do this properly: https://github.com/usnistgov/REFPROP-tests/blob/master/include/REFPROPtests/baseclasses.h#L99-L125

**Comment by Yaykhel** [comment]
Thanks. I used similar approach for REFPROPDll. But I think my problem is with Load_REFPROP argument "path". Loading is failed and it didn't go to SETPATH.

**Comment by Yaykhel** [comment]
When I printed the full library path in RP_join_path(const std::string &one, const std::string &two) in the header.
I noticed the prepend instead of append like this "/librefprop.so/home/yaykhel/Documents/cppKT/cppLibraries/REFPROP"

I further noticed that the "/" is running here and there, too.

**Comment by Yaykhel** [comment]
I fixed the problem. I copied the txt file created in Windows and read in Linux with getline() which carries "CR + LF" (https://stackoverflow.com/questions/45956271/stdgetline-reads-carriage-return-r-into-the-string-how-to-avoid-that). I recreated the txt file in Linux and the program works perfectly. Thanks for your support.

**Comment by ianhbell** [comment]
What txt file are you talking about?

**Comment by ianhbell** [comment]
Also, ``RP_join_path`` appears to be working properly, I think you passed in the arguments backwards.  As usual, without your code it is impossible to know...

**Comment by Yaykhel** [comment]
I included the information of path to REFPROP, refrigerant composition, file names to write after the calculation in a txt file. The program reads the file and load REFPROP. The objective is not to recompile the program every time the refrigerant was changed. The txt file was created on Windows (in fact all codes) and later tested on Linux. Definitely, I didn't misplace arguments otherwise it won't work in Windows. Your header works fine. I diagnosed the problem which is the txt file I created in Windows. After making a new txt file in Linux, the program loaded REFPROP successfully. I think it is a good experience.
